src/operator/tensor/la_op_inline.h

```diff
@@ -0,0 +1,335 @@
+/*!
+ * Copyright (c) 2017 by Contributors
+ * \file la_op_inline.h
+ * \brief Operators for advanced linear algebra.
+ */
+#ifndef MXNET_OPERATOR_TENSOR_LA_OP_INLINE_H_
+#define MXNET_OPERATOR_TENSOR_LA_OP_INLINE_H_
+
+#include <mxnet/c_lapack_api.h>
+
+namespace mxnet {
+namespace op {
+
+using namespace mshadow;
+
+#define LA_OP_NOT_AVAIL " operator can only be called with float/double data type."
+
+// Signature for single matrix operations (decomposition/inversion).
+#define FUNC_SIGNATURE_1(fname, arg1) {CHECK_EQ(MXNET_LAPACK_##fname(MXNET_LAPACK_ROW_MAJOR, 'L', \
+  arg1.size(0), arg1.dptr_, arg1.size(0)), 0) << "fname failed in lapack";}
+
+// Signature for matrix-matrix multiplications involving one diagonal matrix.
+#define FUNC_SIGNATURE_2(fname, arg1, arg2) \
+  { cblas_##fname(CblasRowMajor, (rightside ? CblasRight : CblasLeft), \
+                  CblasLower, (transpose ? CblasTrans : CblasNoTrans), \
+                  CblasNonUnit, arg2.size(0), arg2.size(1), alpha, arg1.dptr_, \
+                  (rightside ? arg2.size(1) : arg2.size(0)), arg2.dptr_, arg2.size(1)); }
+
+
+// Helper functions.
+template<typename DType>
+void CopyLowerToUpper(DType *dptr, int N)
+  { for (int i = 1; i < N; ++i ) for ( int j = 0; j < i; ++j ) dptr[j*N+i] = dptr[i*N+j]; }
+template<typename DType>
+void ZeroUpper(DType *dptr, int N)
+  { for (int i = 0; i < N; ++i ) for ( int j = i+1; j < N; ++j ) dptr[i*N+j] = 0; }
+
+// Forward operators
+
+// D = gemm(A,B,C)
+struct gemm {
+  template<typename xpu, typename DType>
```

Comment:
Due to the nature of the usual operators structure, at the end there must be core operator code that can be compiled with any of the supported types (though it may not be able to run with all of them). I.e. it must compile even when DType is 16-bit, int, etc. But the lapack/blas3 code only supports 32/64-bit and will likely never support anything else. Template specialization with structs allow me to write a default-implementation (that would just error out if it would ever be called) and specialization for 32/64-bits. 
Function overloading would require to write overloads for every potential DType  (as you can not have a default and specialized function templates).